File: dsa/concepts/08_queues/python/a02_queue_using_linked_list.py
"""
Idea:
Implement a queue using a singly linked list.
Maintain two pointers:
- front: for dequeue
- rear: for enqueue
Both operations work in O(1) time.
"""

class ListNode:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


class QueueUsingLinkedList:
    def __init__(self):
        self.front = None
        self.rear = None

    def enqueue(self, data):
        node = ListNode(data)

        if self.front is None and self.rear is None:
            self.front = self.rear = node
        else:
            self.rear.next = node
            self.rear = node

    def dequeue(self):
        if self.is_empty():
            return -1

        data = self.front.val

        if self.front == self.rear:
            self.front = self.rear = None
        else:
            self.front = self.front.next

        return data

    def is_empty(self):
        return self.front is None

    def print_queue(self):
        temp = self.front
        while temp:
            print(temp.val, end=" ")
            temp = temp.next
        print()


# Example usage
if __name__ == "__main__":
    queue = QueueUsingLinkedList()

    queue.enqueue(10)
    queue.enqueue(20)
    print("\nElements:")
    queue.print_queue()

    queue.dequeue()
    print("\nElements:")
    queue.print_queue()

    queue.dequeue()
    print("\nElements:")
    queue.print_queue()

    print(queue.is_empty())

File: dsa/topics/08_queues/java/A02_QueueUsingLinkedList.java
import com.algos.topics.models.ListNode;

public class A02_QueueUsingLinkedList {

	private ListNode front;
	private ListNode rear;
	A02_QueueUsingLinkedList(){
		front = null;
		rear = null;
	}

	public void enqueue(int data){
		ListNode node = new ListNode(data);
		if(front == null && rear == null){
			front = rear = node;
		}else{
			rear.setNext(node);
			rear = node;
		}
	}

	public int dequeue(){
		if(isEmpty())
			return -1;

		int data = front.val;
		if(front == rear){
			front = rear = null;
		}else{
			front = front.getNext();
		}

		return data;
	}
	public boolean isEmpty(){
		if(front == null)
			return true;
		return false;
	}
	public void print(){
		ListNode temp = front;
		while(temp != null){
			System.out.print(temp.getVal()+" ");
			temp = temp.getNext();
		}
	}

	public static void main(String[] args) {
		A02_QueueUsingLinkedList queue = new A02_QueueUsingLinkedList();

		queue.enqueue(10);
		queue.enqueue(20);
		System.out.println("\nelements:");
		queue.print();
		queue.dequeue();
		System.out.println("\nelements:");
		queue.print();
		queue.dequeue();
		System.out.println("\nelements:");
		queue.print();
		System.out.println(queue.isEmpty());
	}
}
